packages/next/src/lib/metadata/resolve-metadata.tsx

```diff
@@ -0,0 +1,243 @@
+import type {
+  Metadata,
+  ResolvedMetadata,
+  ResolvingMetadata,
+} from './types/metadata-interface'
+import type { AbsoluteTemplateString } from './types/metadata-types'
+
+import React from 'react'
+import { createDefaultMetadata } from './constant'
+import { resolveOpenGraph } from './resolve-opengraph'
+import { resolveTitle } from './resolve-title'
+import { elementsFromResolvedOpenGraph } from './generate/opengraph'
+import { elementsFromResolvedBasic } from './generate/basic'
+import { elementsFromResolvedAlternates } from './generate/alternate'
+
+type Item =
+  | {
+      type: 'layout' | 'page'
+      // A number that represents which layer the item is in. Starting from 0.
+      layer: number
+      mod: () => Promise<{
+        metadata?: Metadata
+        generateMetadata?: (
+          props: any,
+          parent: ResolvingMetadata
+        ) => Promise<Metadata>
+      }>
+      path: string
+    }
+  | {
+      type: 'icon'
+      // A number that represents which layer the item is in. Starting from 0.
+      layer: number
+      mod?: () => Promise<{
+        metadata?: Metadata
+        generateMetadata?: (
+          props: any,
+          parent: ResolvingMetadata
+        ) => Promise<Metadata>
+      }>
+      path?: string
+    }
+
+// Merge the source metadata into the resolved target metadata.
+function merge(
+  target: ResolvedMetadata,
+  source: Metadata,
+  templateStrings: {
+    title: AbsoluteTemplateString
+    openGraphTitle: AbsoluteTemplateString
+    twitterTitle: AbsoluteTemplateString
```

Comment:
just pass the templates here, we don't need the whole string object, these shouldn't be mutated and we don't need the absolute title for anything